src/runner.py

```diff
@@ -0,0 +1,50 @@
+import argparse
+from solver import Solver
+from utils import render_grid
+from model import Grid
+
+
+def run():
+    parser = argparse.ArgumentParser(
+        description="A Sudoku solver which applies solving rules just as a human would."
+    )
+    parser.add_argument("input", help="The sudoku puzzle to solve.")
+
+    args = parser.parse_args()
+
+    with open(args.input, "r") as f:
+        input_lines = f.readlines()
+        grid = Grid.from_rows_notation(input_lines)
+
+    render_grid(grid).show()
+    print(apply_solver(grid))
+    render_grid(grid).show()
+
+
+def apply_solver(grid: Grid) -> str:
```

Comment:
Extracting the application of the Solver class allows us to unit test the way the Solver is applied without getting involved with IO operations on input files.